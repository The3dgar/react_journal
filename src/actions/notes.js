import Swal from "sweetalert2";
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from "../helpers/fileUpload";
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";

export const startNewNotes = () => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    const newNote = {
      title: "",
      body: "",
      date: new Date().getTime(),
    };

    const doc = await db.collection(`${uid}/journal/notes`).add(newNote);
    dispatch(activeNote(doc.id, newNote));
    dispatch(addNewNote(doc.id, newNote))
  };
};

export const addNewNote = (id, note) => ({
  type: types.notesAddNew,
  payload: {
    id, note
  }
})

export const activeNote = (id, note) => ({
  type: types.notesActive,
  payload: {
    id,
    ...note,
  },
});

export const setNotes = (notes) => ({
  type: types.notesLoad,
  payload: notes,
});

export const startLoadingNotes = (uid) => {
  return async (dispatch) => {
    const notes = await loadNotes(uid);
    dispatch(setNotes(notes));
  };
};

export const startSaveNote = (note) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;

    if (!note.url) {
      delete note.url;
    }
    const noteToFirestore = { ...note };
    delete noteToFirestore.id;

    await db
      .doc(`${uid}/journal/notes/${note.id}`)
      .update(noteToFirestore)
      .catch((e) => {
        console.log(e);
        Swal.fire("error en la nota", note.title, "error");
      });

    dispatch(refreshNote(note.id, noteToFirestore));

    Swal.fire("Saved", note.title, "success");
  };
};

export const refreshNote = (id, note) => ({
  type: types.notesUpdate,
  payload: {
    id,
    note: {
      id,
      ...note,
    },
  },
});

export const startUpLoading = (file) => {
  return async (dispatch, getState) => {
    const { active: activeNote } = getState().notes;
    Swal.fire({
      title: "Uploading file...",
      text: "Please wait",
      allowOutSideClick: false,
      onBeforeOpen: () => Swal.showLoading(),
    });
    const url = await fileUpload(file).catch((e) =>
      console.log("ERROR IN FETCH CLOUDINARY: ", e.toString())
    );
    activeNote.url = url;
    dispatch(startSaveNote(activeNote));

    Swal.close();
  };
};

export const startDeleting = (id) => {
  return async (dispatch, getState) => {
    const { uid } = getState().auth;
    await db.doc(`${uid}/journal/notes/${id}`).delete()

    dispatch(deleteNote(id))
  };
};

export const deleteNote = id => ({
  type : types.notesDelete,
  payload: id
})

export const notesLogout = () => ({
  type: types.notesLogoutCleaning,
})