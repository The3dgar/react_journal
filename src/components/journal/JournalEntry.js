import React from "react";

export const JournalEntry = () => {
  return (
    <div className="journal__entry pointer">
      <div
        className="journal__entry-picture"
        style={{
          backgroundSize: "cover",
          backgroundImage:
            "url(https://haciendofotos.com/wp-content/uploads/las-mejores-fotos-de-paisajes-2020.jpg)",
        }}
      ></div>

      <div className="journal__entry-body">
        <p className="journal__entry-title"> El karma bueno</p>
        <p className="journal__entry-content">
          loremadsadsadsadada adsoia adsadmiaod askdnmaods
        </p>
      </div>

      <div className="journal__entry-date-box">
        <span>Monday</span>
        <h4>28</h4>
      </div>
    </div>
  );
};
