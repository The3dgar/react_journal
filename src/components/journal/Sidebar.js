import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startLogout } from "../../actions/auth";
import { startNewNotes } from "../../actions/notes";
import { JournalEntries } from "./JournalEntries";

export const Sidebar = () => {
  const dispatch = useDispatch();
  const { name } = useSelector(state => state.auth)


  const handleLogout = () => {
    dispatch(startLogout())
  };

  const handleAddNote = () => {
    dispatch(startNewNotes())
  }
  return (
    <aside className="journal__sidebar-content">
      <div className="journal__sidebar-navbar">
        <h3 className="mt-5">
          <i className="far fa-moon" />
          <span> {name}</span>
        </h3>

        <button className="btn" onClick={handleLogout}>
          Logout
        </button>
      </div>

      <div className="journal__new-entry" onClick={handleAddNote}>
        <i className="far fa-calendar-plus fa-5x" />
        <p className="mt-5">New entry</p>
      </div>

      <JournalEntries />
    </aside>
  );
};
