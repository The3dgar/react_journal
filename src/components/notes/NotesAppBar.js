import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { startSaveNote, startUpLoading } from "../../actions/notes";

export const NotesAppBar = () => {
  const { active } = useSelector((state) => state.notes);

  const handleSave = () => {
    dispatch(startSaveNote(active));
  };

  const handlePictureSave = () => {
    document.querySelector("#fileSelector").click()
  };
  const dispatch = useDispatch()

  const handleFileChange = (e) => {
    const file = e.target.files[0]
    if(file){
      dispatch(startUpLoading(file))
    }
  };

  const today = () => new Date().toDateString()

  return (
    <div className="notes_appbar">
      <span>{today()}</span>
      <input
        id="fileSelector"
        type="file"
        style={{ display: "none" }}
        onChange={handleFileChange}
      />
      <div>
        <button className="btn" onClick={handlePictureSave}>
          Picture
        </button>
        <button className="btn" onClick={handleSave}>
          Save
        </button>
      </div>
    </div>
  );
};
