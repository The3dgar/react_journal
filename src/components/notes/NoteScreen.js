import React from 'react'
import { NotesAppBar } from './NotesAppBar'

export const NoteScreen = () => {
  return (
    <div className="notes__main-content">
      <NotesAppBar/>
      <div className="notes__content">
        <input type="text" placeholder="Title" className="notes__title-input" autoComplete="off"/>
        <textarea placeholder="Today was..." className="notes__textarea"></textarea>
        <div className="notes__image">
          <img src="https://www.tom-archer.com/wp-content/uploads/2018/06/milford-sound-night-fine-art-photography-new-zealand.jpg" alt="land"/>
        </div>
      </div>
    </div>
  )
}
