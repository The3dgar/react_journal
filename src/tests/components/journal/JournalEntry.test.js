import { mount } from "enzyme";
import React from "react";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import { JournalEntry } from "../../../components/journal/JournalEntry";
import { activeNote } from "../../../actions/notes";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initState = {};
let store = mockStore(initState);
store.dispatch = jest.fn();

const nota = {
  id:10,
  date: 0,
  title: "Titulo",
  body:"mundo",
  url: "https://algunhost.com/picture.jpg"
}

const wrapper = mount(
  <Provider store={store}>
    <JournalEntry {...nota}/>
  </Provider>
);

describe("test in <JournalEntr/>", () => {
  test("should snap correctly", () => {
    expect(wrapper).toMatchSnapshot()
  });

  test('should dispatch active note', async () => {
    wrapper.find(".journal__entry").prop("onClick")()

    expect(store.dispatch).toHaveBeenCalledWith(activeNote(nota.id, {...nota}))
    

  })
  
});
