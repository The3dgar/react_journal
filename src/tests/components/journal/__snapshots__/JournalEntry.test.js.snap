// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test in <JournalEntr/> should snap correctly 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn anime__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://algunhost.com/picture.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
       
      Titulo
    </p>
    <p
      className="journal__entry-content"
    >
      mundo
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31
    </h4>
  </div>
</div>
`;
